shared/src/sources/balancer/graph_api.rs

```diff
@@ -70,18 +75,57 @@ impl BalancerSubgraphClient {
             if let Some(last_pool) = page.last() {
                 last_id = last_pool.id;
             }
-
-            for pool in page {
-                pools_by_factory
-                    .entry(pool.factory.unwrap_or_default())
-                    .or_default()
-                    .push(pool.into_registered(block_number)?);
-            }
-
+            result.extend(page);
             has_next_page
         } {}
+        Ok(result)
+    }
 
-        Ok(RegisteredWeightedPools {
+    /// Retrieves the list of registered pools from the subgraph.
+    pub async fn get_registered_pools(&self) -> Result<RegisteredPools> {
+        let block_number = self.get_safe_block().await?;
+
+        let weighted_pools = self
+            .query_graph_for::<WeightedToken>(block_number, pools_query::WEIGHTED_POOL_QUERY)
+            .await?
+            .into_iter()
+            .map(|pool_data| {
+                (
+                    pool_data.factory,
+                    pool_data.into_registered_pool(block_number),
+                )
+            });
+        let stable_pools = self
+            .query_graph_for::<StableToken>(block_number, pools_query::STABLE_POOL_QUERY)
+            .await?
+            .into_iter()
+            .map(|pool_data| {
+                (
+                    pool_data.factory,
+                    pool_data.into_registered_pool(block_number),
+                )
+            });
+
+        let mut pools_by_factory = HashMap::<H160, Vec<RegisteredPool>>::new();
+        for (factory, pool) in weighted_pools.chain(stable_pools) {
+            let pool = match pool {
+                Ok(pool) => pool,
+                Err(err) => {
+                    // Technically this should never happen and would only ever be from
+                    // a token with more than 18 decimals (not supported by balancer contracts)
+                    // https://github.com/balancer-labs/balancer-v2-monorepo/blob/deployments-latest/pkg/pool-utils/contracts/BasePool.sol#L476-L487
+                    // We, thus, log but skip this error.
+                    tracing::error!("Unexpected failure to build registered pool {}", err);
+                    continue;
```

Comment:
If this should never happen, then I would return an error here. I would argue that if this were to happen, we would want the pool initialization to fail, as it would mean that there is something incorrect about our assumptions of how the contracts/subgraph work

Reply:
Ben and I discussed this in the previous bigger PR. I suggested we shouldn't fail the whole process. But I can see your argument too. Either way is fine with me.